packages/protocol/test/tokenomics/blockFee.test.ts

```diff
@@ -0,0 +1,135 @@
+            proposerSigner,
+            genesisHeight,
+            genesisHash,
+            tkoTokenL1,
+            l1AddressManager,
+            interval,
+        } = await initTokenomicsFixture());
+    });
+
+    afterEach(() => clearInterval(interval));
+
+    it("expects the blockFee to go be 0 when no periods have passed", async function () {
+        // deploy a new instance of TaikoL1 so no blocks have passed.
+        const tL1 = await deployTaikoL1(l1AddressManager, genesisHash, true);
+        const blockFee = await tL1.getBlockFee();
+        expect(blockFee.eq(0)).to.be.eq(true);
+    });
+
+    it("block fee should increase as the halving period passes, while no blocks are proposed", async function () {
+        const { bootstrapDiscountHalvingPeriod } = await taikoL1.getConfig();
+
+        const iterations: number = 5;
+        const period: number = bootstrapDiscountHalvingPeriod
+            .mul(1000)
+            .toNumber();
+
+        let lastBlockFee: BigNumber = await taikoL1.getBlockFee();
+
+        for (let i = 0; i < iterations; i++) {
+            await sleep(period);
+            const blockFee = await taikoL1.getBlockFee();
+            expect(blockFee.gt(lastBlockFee)).to.be.eq(true);
+            lastBlockFee = blockFee;
+        }
+    });
+
+    it("proposes blocks on interval, blockFee should increase, proposer's balance for TKOToken should decrease as it pays proposer fee, proofReward should increase since slots are growing and no proofs have been submitted", async function () {
+        const { maxNumBlocks, commitConfirmations } = await taikoL1.getConfig();
+        // wait for one period of halving to occur, so fee is not 0.
+        const blockIdsToNumber: any = {};
+
+        // set up a proposer to continually propose new blocks
+        const proposer = new Proposer(
+            taikoL1.connect(proposerSigner),
+            l2Provider,
+            commitConfirmations.toNumber(),
+            maxNumBlocks.toNumber(),
+            0,
+            proposerSigner
+        );
+
+        // get the initiaal tkoBalance, which should decrease every block proposal
+        let lastProposerTkoBalance = await tkoTokenL1.balanceOf(
+            await proposerSigner.getAddress()
+        );
+
+        // do the same for the blockFee, which should increase every block proposal
+        // with proofs not being submitted.
+        // we want to wait for enough blocks until the blockFee is no longer 0, then run our
+        // tests.
+        let lastBlockFee = await taikoL1.getBlockFee();
+        while (lastBlockFee.eq(0)) {
+            await sleep(500);
+            lastBlockFee = await taikoL1.getBlockFee();
+        }
+
+        let lastProofReward = BigNumber.from(0);
+
+        let hasFailedAssertions: boolean = false;
+        let blocksProposed: number = 0;
+        // every time a l2 block is created, we should try to propose it on L1.
+        l2Provider.on("block", async (blockNumber: number) => {
```

Comment:
Shall we wrap the event emitter into a [JavaScript generator function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*#generator_example)? and then we can handle everything in the main javascript event loop like:
```js
function* newL2BlockGenerator() {
            l2Provider.on("block", (blockNumber: number) => yield blockNumber);
}

for (const blockNumber of newL2BlockGenerator()) {
            // ...
            // handle everything including error in the main Node.js event loop
}
```
which may make the error assertion logic more simpler and the following `await sleep()` code unnecessary?

Reply:
Sounds good, I can try this tomorrow

Reply:
Rather than do the generator method, which seemed kind of clunky for this (somehow I had difficulty yielding), i did a custom EventEmitter. can you take a look and see if thats OK? I also moved the listeners to their own utils methods where reusable.

Reply:
You can't actually use `yield` from the event listener itself. I refactored to some custom events and eventmitter lib, and its a lot cleaner now. Removed most of the await sleep code. Unsure if theres still a way you want it doen with the Generator. Take a looksee and lemme know?